File: CSharp/13-copilot-hooks/NotificationService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Demo.Services;

public enum NotificationChannel { Email, Sms, Push }

public enum NotificationStatus { Pending, Sent, Failed, Suppressed }

public record Notification(
    Guid Id,
    string RecipientId,
    NotificationChannel Channel,
    string Subject,
    string Body,
    NotificationStatus Status,
    DateTime CreatedAt,
    DateTime? SentAt);

public record SendNotificationRequest(
    string RecipientId,
    NotificationChannel Channel,
    string Subject,
    string Body);

public class NotificationService
{
    private readonly Dictionary<Guid, Notification> _store = new();
    private readonly HashSet<string> _suppressedRecipients = new();

    public async Task<Notification> SendAsync(
        SendNotificationRequest request,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(request.RecipientId))
            throw new ArgumentException("RecipientId cannot be empty.", nameof(request));
        if (string.IsNullOrWhiteSpace(request.Subject))
            throw new ArgumentException("Subject cannot be empty.", nameof(request));
        if (string.IsNullOrWhiteSpace(request.Body))
            throw new ArgumentException("Body cannot be empty.", nameof(request));

        await Task.Delay(0, cancellationToken);

        var status = _suppressedRecipients.Contains(request.RecipientId)
            ? NotificationStatus.Suppressed
            : NotificationStatus.Sent;

        var notification = new Notification(
            Guid.NewGuid(),
            request.RecipientId,
            request.Channel,
            request.Subject.Trim(),
            request.Body.Trim(),
            status,
            DateTime.UtcNow,
            status == NotificationStatus.Sent ? DateTime.UtcNow : null);

        _store[notification.Id] = notification;
        return notification;
    }

    public async Task<Notification?> GetByIdAsync(
        Guid notificationId,
        CancellationToken cancellationToken = default)
    {
        await Task.Delay(0, cancellationToken);
        _store.TryGetValue(notificationId, out var notification);
        return notification;
    }

    public async Task<IReadOnlyList<Notification>> GetForRecipientAsync(
        string recipientId,
        NotificationChannel? channel = null,
        int page = 1,
        int pageSize = 20,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(recipientId))
            throw new ArgumentException("RecipientId cannot be empty.", nameof(recipientId));
        if (page < 1)
            throw new ArgumentOutOfRangeException(nameof(page));
        if (pageSize is < 1 or > 100)
            throw new ArgumentOutOfRangeException(nameof(pageSize));

        await Task.Delay(0, cancellationToken);

        return _store.Values
            .Where(n => n.RecipientId == recipientId
                     && (channel == null || n.Channel == channel))
            .OrderByDescending(n => n.CreatedAt)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToList()
            .AsReadOnly();
    }

    public void SuppressRecipient(string recipientId)
    {
        if (string.IsNullOrWhiteSpace(recipientId))
            throw new ArgumentException("RecipientId cannot be empty.", nameof(recipientId));
        _suppressedRecipients.Add(recipientId);
    }

    public void UnsuppressRecipient(string recipientId) =>
        _suppressedRecipients.Remove(recipientId);

    public bool IsRecipientSuppressed(string recipientId) =>
        _suppressedRecipients.Contains(recipientId);

    public int TotalCount => _store.Count;

    public int SentCount => _store.Values.Count(n => n.Status == NotificationStatus.Sent);
}

File: Java/13-copilot-hooks/NotificationService.java
import java.time.Instant;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class NotificationService {

    public enum Channel { EMAIL, SMS, PUSH }

    public enum Status { PENDING, SENT, FAILED, SUPPRESSED }

    public record Notification(
            UUID id,
            String recipientId,
            Channel channel,
            String subject,
            String body,
            Status status,
            Instant createdAt,
            Instant sentAt) {}

    public record SendNotificationRequest(
            String recipientId,
            Channel channel,
            String subject,
            String body) {}

    private final Map<UUID, Notification> store = new ConcurrentHashMap<>();
    private final Set<String> suppressedRecipients = ConcurrentHashMap.newKeySet();

    public Notification send(SendNotificationRequest request) {
        if (request.recipientId() == null || request.recipientId().isBlank())
            throw new IllegalArgumentException("recipientId cannot be empty.");
        if (request.subject() == null || request.subject().isBlank())
            throw new IllegalArgumentException("subject cannot be empty.");
        if (request.body() == null || request.body().isBlank())
            throw new IllegalArgumentException("body cannot be empty.");
        if (request.channel() == null)
            throw new IllegalArgumentException("channel cannot be null.");

        Status status = suppressedRecipients.contains(request.recipientId())
                ? Status.SUPPRESSED
                : Status.SENT;

        Instant now = Instant.now();

        Notification notification = new Notification(
                UUID.randomUUID(),
                request.recipientId().trim(),
                request.channel(),
                request.subject().trim(),
                request.body().trim(),
                status,
                now,
                status == Status.SENT ? now : null);

        store.put(notification.id(), notification);
        return notification;
    }

    public Optional<Notification> getById(UUID notificationId) {
        return Optional.ofNullable(store.get(notificationId));
    }

    public List<Notification> getForRecipient(String recipientId, Channel channel, int page, int pageSize) {
        if (recipientId == null || recipientId.isBlank())
            throw new IllegalArgumentException("recipientId cannot be empty.");
        if (page < 1)
            throw new IllegalArgumentException("page must be >= 1.");
        if (pageSize < 1 || pageSize > 100)
            throw new IllegalArgumentException("pageSize must be between 1 and 100.");

        return store.values().stream()
                .filter(n -> n.recipientId().equals(recipientId))
                .filter(n -> channel == null || n.channel() == channel)
                .sorted(Comparator.comparing(Notification::createdAt).reversed())
                .skip((long) (page - 1) * pageSize)
                .limit(pageSize)
                .collect(Collectors.toUnmodifiableList());
    }

    public void suppressRecipient(String recipientId) {
        if (recipientId == null || recipientId.isBlank())
            throw new IllegalArgumentException("recipientId cannot be empty.");
        suppressedRecipients.add(recipientId);
    }

    public void unsuppressRecipient(String recipientId) {
        suppressedRecipients.remove(recipientId);
    }

    public boolean isRecipientSuppressed(String recipientId) {
        return suppressedRecipients.contains(recipientId);
    }

    public int getTotalCount() {
        return store.size();
    }

    public long getSentCount() {
        return store.values().stream()
                .filter(n -> n.status() == Status.SENT)
                .count();
    }
}
